File: mixed-persistence/mixed-persistence-scala-sbt/hello-impl/src/main/scala/com/lightbend/lagom/samples/hello/impl/HelloServiceImpl.scala
package com.lightbend.lagom.samples.hello.impl

import akka.Done
import akka.NotUsed
import akka.cluster.sharding.typed.scaladsl.ClusterSharding
import akka.cluster.sharding.typed.scaladsl.EntityRef
import akka.util.Timeout
import com.lightbend.lagom.samples.hello.api.{ Greeting, HelloService }
import com.lightbend.lagom.samples.hello.impl.readside.{ GreetingsRepository, ReadSideGreeting }
import com.lightbend.lagom.scaladsl.api.ServiceCall
import com.lightbend.lagom.scaladsl.api.transport.BadRequest
import com.lightbend.lagom.scaladsl.persistence.PersistentEntityRegistry

import scala.concurrent.ExecutionContext
import scala.concurrent.duration._
import com.lightbend.lagom.samples.hello.api


class HelloServiceImpl(
  greetingsRepository: GreetingsRepository,
  clusterSharding: ClusterSharding,
  persistentEntityRegistry: PersistentEntityRegistry
)(implicit ec: ExecutionContext)
    extends HelloService {

  /**
    * Looks up the entity for the given ID.
    */
  private def entityRef(id: String): EntityRef[HelloCommand] =
    clusterSharding.entityRefFor(HelloState.typeKey, id)

  implicit val timeout = Timeout(5.seconds)

  override def hello(id: String): ServiceCall[NotUsed, String] = ServiceCall {
    _ =>
      // Look up the sharded entity (aka the aggregate instance) for the given ID.
      val ref = entityRef(id)

      // Ask the aggregate instance the Hello command.
      ref
        .ask[Greeting](replyTo => Hello(id, replyTo))
        .map(greeting => greeting.message)
  }

  override def useGreeting(id: String) = ServiceCall { request =>
    // Look up the sharded entity (aka the aggregate instance) for the given ID.
    val ref = entityRef(id)

    // Tell the aggregate to use the greeting message specified.
    ref
      .ask[Confirmation](
        replyTo => UseGreetingMessage(request.message, replyTo)
      )
      .map {
        case Accepted => Done
        case _        => throw BadRequest("Can't upgrade the greeting message.")
      }
  }

  override def allGreetings(): ServiceCall[NotUsed, Seq[api.Greeting]] =
    ServiceCall { _ =>
      greetingsRepository
        .getAll()
        .map(_.map(toApi).toSeq)
    }

  private def toApi: ReadSideGreeting => api.Greeting = { readSideModel =>
    api.Greeting(readSideModel.name, readSideModel.message)
  }
}

File: mixed-persistence/mixed-persistence-java-sbt/hello-impl/src/main/java/com/lightbend/lagom/samples/mixedpersistence/hello/impl/HelloServiceImpl.java
package com.lightbend.lagom.samples.mixedpersistence.hello.impl;

import akka.Done;
import akka.NotUsed;
import akka.cluster.sharding.typed.javadsl.ClusterSharding;
import akka.cluster.sharding.typed.javadsl.Entity;
import akka.cluster.sharding.typed.javadsl.EntityRef;
import com.lightbend.lagom.javadsl.api.ServiceCall;
import com.lightbend.lagom.javadsl.api.transport.BadRequest;
import com.lightbend.lagom.samples.mixedpersistence.hello.api.GreetingMessage;
import com.lightbend.lagom.samples.mixedpersistence.hello.api.HelloService;
import com.lightbend.lagom.samples.mixedpersistence.hello.api.UserGreeting;
import com.lightbend.lagom.samples.mixedpersistence.hello.impl.entity.HelloAggregate;
import com.lightbend.lagom.samples.mixedpersistence.hello.impl.entity.HelloCommand;
import com.lightbend.lagom.samples.mixedpersistence.hello.impl.entity.HelloCommand.Hello;
import com.lightbend.lagom.samples.mixedpersistence.hello.impl.entity.HelloCommand.UseGreetingMessage;
import com.lightbend.lagom.samples.mixedpersistence.hello.impl.readside.Greetings;
import org.pcollections.PSequence;

import javax.inject.Inject;
import java.time.Duration;

/**
 * Implementation of the HelloService.
 */
public class HelloServiceImpl implements HelloService {
    private final Greetings greetings;
    private final Duration askTimeout = Duration.ofSeconds(5);
    private ClusterSharding clusterSharding;

    @Inject
    public HelloServiceImpl(Greetings greetings, ClusterSharding clusterSharding) {
        this.greetings = greetings;
        this.clusterSharding = clusterSharding;
        // register the Aggregate as a sharded entity
        this.clusterSharding.init(
            Entity.of(
                HelloAggregate.ENTITY_TYPE_KEY,
                HelloAggregate::create
            )
        );
    }

    @Override
    public ServiceCall<NotUsed, String> hello(String id) {
        return request -> {
            // Look up the aggregete instance for the given ID.
            EntityRef<HelloCommand> ref = clusterSharding.entityRefFor(HelloAggregate.ENTITY_TYPE_KEY, id);
            // Ask the entity the Hello command.
            return ref.
                <HelloCommand.Greeting>ask(replyTo -> new Hello(id, replyTo), askTimeout)
                .thenApply(greeting -> greeting.message);
        };
    }

    @Override
    public ServiceCall<GreetingMessage, Done> useGreeting(String id) {
        return request -> {
            // Look up the aggregete instance for the given ID.
            EntityRef<HelloCommand> ref = clusterSharding.entityRefFor(HelloAggregate.ENTITY_TYPE_KEY, id);
            // Tell the entity to use the greeting message specified.
            return ref.
                <HelloCommand.Confirmation>ask(replyTo -> new UseGreetingMessage(request.getMessage(), replyTo), askTimeout)
                .thenApply(confirmation -> {
                        if (confirmation instanceof HelloCommand.Accepted) {
                            return Done.getInstance();
                        } else {
                            throw new BadRequest(((HelloCommand.Rejected) confirmation).reason);
                        }
                    }
                );
        };
    }


    @Override
    public ServiceCall<NotUsed, PSequence<UserGreeting>> allGreetings() {
        return request -> greetings.all();
    }
}
